src/Network/Transport/TCP.hs

```diff
@@ -962,35 +1005,45 @@ handleConnectionRequest transport socketClosed (sock, sockAddr) = handle handleE
       (actualHost, actualPort) <-
         decodeSockAddr sockAddr >>=
           maybe (throwIO (userError "handleConnectionRequest: invalid socket address")) return
-      (ourEndPointId, theirAddress) <- do
+      -- The peer must send our identifier and their address promptly, if a
+      -- timeout is set.
+      addrInfo <- do
         ourEndPointId <- recvWord32 sock
         let maxAddressLength = tcpMaxAddressLength $ transportParams transport
-        theirAddress <- EndPointAddress . BS.concat <$>
-          recvWithLength maxAddressLength sock
-        return (ourEndPointId, theirAddress)
-      let ourAddress = encodeEndPointAddress (transportHost transport)
-                                             (transportPort transport)
-                                             ourEndPointId
-      (theirHost, _, _)
-        <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
-                 return
-                 (decodeEndPointAddress theirAddress)
+        mTheirAddress <- BS.concat <$> recvWithLength maxAddressLength sock
+        -- Sending a length = 0 address means unaddressable.
+        if BS.null mTheirAddress
+        then fmap ((,) ourEndPointId . Left) randomEndPointAddress
+        else do
+          let theirAddress = EndPointAddress mTheirAddress
+          (theirHost, theirPort, theirEndPointId)
+            <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
+                     return
+                     (decodeEndPointAddress theirAddress)
+          return (ourEndPointId, Right (theirAddress, (theirHost, theirPort, theirEndPointId)))
+      (ourEndPointId, theirAddress, peerHost) <- case addrInfo of
+        (x, Left y) -> return (x, y, Nothing)
+        (x, Right (y, (h, _, _))) -> return (x, y, Just h)
```

Comment:
Can this case analysis be eliminated by having the closure producing addrInfo produce ` (ourEndPointId, theirAddress, peerHost)` instead.